### PR TITLE
Add providerCode & courseCode params flow for existing users

### PR DESCRIPTION
**Pull request by davidgisbey**
## Context

This PR builds on the work done in #1165 

It implements the flow for existing candidates who arrive from find with course params in the query string.

The logic for the expected is behaviour is similar to that of a new Candidate. 

## Changes proposed in this pull request
The desired behaviour is

If they have less than 3 course options && course_from_find_id is present on a candidate && and the course has one site

  - It should add an ApplicationChoice with the relevant CourseOption. 
  - Set the candidates course_from_find_id to nil
  - Redirect to the candidate_interface_course_choices_review_path 

elsif they have less than 3 course option && course_from_find_id is present on a candidate && the course has multiple sites
  - It should redirect the site selection page for their selected course
  - set the candidates course_from_find_id to nil

elsif they already have 3 course choices
    - Redirect to the candidate_interface_course_choices_review_path 
    - Present a flash message telling them they can't have more than 3 course options

image

## Guidance to review
- Navigate the login flow with a course with one site (/candidate/apply?providerCode=C58&courseCode=3D2X)

- Navigate the login flow with a course with multiple sites (/candidate/apply?providerCode=2LR&courseCode=3255)

- Add any 3 courses while logged in. Logout, then follow the flow for any other course in the Apply database.

## Link to Trello card

https://trello.com/c/EQQoWqun/632-you-selected-a-course-page-a-candidate-can-confirm-their-course-choice-if-they-have-come-from-find

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/controllers/candidate_interface/sign_up_controller.rb [hunk 1]
```diff
@@ -14,6 +14,8 @@ def create
       if @sign_up_form.existing_candidate?
         MagicLinkSignIn.call(candidate: @sign_up_form.candidate)
         add_identity_to_log @sign_up_form.candidate.id
+        candidate = Candidate.find(@sign_up_form.candidate.id)
+        candidate.update(course_from_find_id: @sign_up_form.course_from_find_id)
```
Comment (tijmenb) [inline app/controllers/candidate_interface/sign_up_controller.rb]:
```suggestion
        candidate.update!(course_from_find_id: @sign_up_form.course_from_find_id)
```

File: spec/services/existing_candidate_authentication_spec.rb [hunk 2]
```diff
@@ -11,8 +11,26 @@
...
+    context 'when the candidate already has 3 application choices' do
+      it 'sets the candidates course_from_find_id to nil and returns :candidate_already_has_3_courses' do
+        new_course = create(:course)
+        candidate = create(:candidate, course_from_find_id: new_course.id)
+        provider = create(:provider)
+        application_form = create(:application_form, candidate: candidate)
+        3.times { course_option_for_provider(provider: provider) }
+        provider.courses.each do |course|
+          create(:application_choice, application_form: application_form, course_option_id: course.course_options.first.id)
+        end
```
Comment (tijmenb) [inline spec/services/existing_candidate_authentication_spec.rb]: The 3 courses don't have to be from the same provider, so perhaps these lines can be shortened to:

```
create(:completed_application_form, candidate: candidate, application_choices_count: 3)
```

Comment (davidgisbey) [reply spec/services/existing_candidate_authentication_spec.rb]: Nice. Have done. Thanks 👍

**Comment by davidgisbey** [comment]
Have rewritten the spec to comply with the style guide.

File: spec/services/candidate_interface/existing_candidate_authentication_spec.rb [hunk 3]
```diff
@@ -4,8 +4,25 @@
...
+      it 'sets the candidates course_from_find_id to nil and sets candidate_already_has_3_courses too true' do
+        new_course = create(:course)
+        candidate = create(:candidate, course_from_find_id: new_course.id)
+        create(:completed_application_form, candidate: candidate, application_choices_count: 3)
+
+        service = described_class.new(candidate: candidate)
+        service.execute
+
+        expect(service).to be_candidate_already_has_3_courses
+        expect(service).not_to be_candidate_has_new_course_added
```
Comment (davidgisbey) [inline spec/services/candidate_interface/existing_candidate_authentication_spec.rb]: Not a big fan of this syntax, but Rubcop was complaining 🤷‍♂

Comment (davidgisbey) [reply spec/services/candidate_interface/existing_candidate_authentication_spec.rb]: I personally find `expect(service.candidate_already_has_3_courses?).to be_truthy`

Comment (tijmenb) [reply spec/services/candidate_interface/existing_candidate_authentication_spec.rb]: Yeah, agreed, "to be candidate has new course added" totally salad like sounds weird a word.

Can you turn off the cop?

Comment (davidgisbey) [reply spec/services/candidate_interface/existing_candidate_authentication_spec.rb]: Have done!

File: spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb [hunk 4]
```diff
@@ -0,0 +1,158 @@
+require 'rails_helper'
+
+RSpec.describe 'An existing candidate arriving from Find with a course and provider code' do
+  include CourseOptionHelpers
+  scenario 'retaining their course selection through the sign up process' do
+    given_the_pilot_is_open
+    and_confirm_course_choice_from_find_is_activated
+    and_i_an_existing_candidate_on_apply
```
Comment (tvararu) [inline spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb]:
```suggestion
    and_i_am_an_existing_candidate_on_apply
```

File: spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    and_i_have_less_than_3_application_options
+    and_the_course_i_selected_only_has_one_site
+
+    when_i_arrive_at_the_sign_up_page_with_course_params_with_one_site
+    and_i_submit_my_email_address
+    and_click_on_the_magic_link
+    then_i_should_see_the_courses_review_page
+    and_i_should_see_the_course_name_and_code
+    and_i_should_see_the_site
+    and_my_find_from_id_course_should_be_set_to_nil
```
Comment (tvararu) [inline spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb]: Nit: doesn't feel like we should test for implementation details in a system spec, but the alternative in this particular case would be to duplicate a bunch of our assertions. We'd need to go through the whole process again, and make sure we don't get another offer to add the course. So I'm ok with this!

File: spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+    given_the_course_i_selected_has_multiple_sites
+    and_i_an_existing_candidate_on_apply
```
Comment (tvararu) [inline spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb]:
```suggestion
    and_i_am_an_existing_candidate_on_apply
```

File: spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+    when_i_arrive_at_the_sign_up_page_with_course_params_with_multiple_sites
+    and_i_submit_my_email_address
+    and_click_on_the_magic_link
+    then_i_should_see_the_course_choices_site_page
+    and_i_see_the_form_to_pick_a_location
+    and_my_find_from_id_course_should_be_set_to_nil
+
+    and_the_course_i_selected_only_has_one_site
+    and_i_an_existing_candidate_on_apply
```
Comment (tvararu) [inline spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb]:
```suggestion
    and_i_am_an_existing_candidate_on_apply
```

File: spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    FeatureFlag.activate('pilot_open')
+  end
+
+  def and_the_course_i_selected_only_has_one_site
+    @course = create(:course, exposed_in_find: true, open_on_apply: true, name: 'Potions')
+    @site = create(:site, provider: @course.provider)
+    create(:course_option, site: @site, course: @course, vacancy_status: 'B')
+  end
+
+  def and_i_an_existing_candidate_on_apply
```
Comment (tvararu) [inline spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb]:
```suggestion
  def and_i_am_an_existing_candidate_on_apply
```

File: spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb (hunk 5 above)
Comment (tvararu) [inline spec/system/candidate_interface/candidate_existing_user_with_course_params_spec.rb]:
```suggestion
    and_my_course_from_find_id_should_be_set_to_nil
```

Maybe?

**Comment by davidgisbey** [comment]
Thank god there are non-native English speakers out there to spell/grammar check me 😂 

Will fix and then merge. Thanks Theo!

**Pull request merged**